pkg/sshd/lock.go

```diff
@@ -26,11 +26,32 @@ type RepositoryLock interface {
 }
 
 func wrapInLock(lck RepositoryLock, repoName string, timeout time.Duration, fn func() error) error {
-	if err := lck.Lock(repoName, timeout); err != nil {
+	if err := lck.Lock(repoName); err != nil {
 		return errAlreadyLocked
 	}
-	defer lck.Unlock(repoName, timeout)
-	return fn()
+	tick := time.Tick(timeout)
+	quit := make(chan bool)
+	var fnerr error
+	go func() {
+		fnerr = fn()
+		quit <- false
+	}()
+	go func() {
+		for {
+			select {
+			case <-tick:
+				quit <- true
```

Comment:
this will block forever if the function succeeds and `quit <- true` (above) executes first